### DISCUSSION TITLE
ModuleBase::Vector3 should support manipulation between int and double.

### DISCUSSION THREAD

**Discussion by ErjieWu** [deepmodeling/abacus-develop#5830] [Note for Developers]
### Background

Currently, when multiplying two vector with type ModuleBase::Vector3<int> and ModuleBase::Vector3<double>, it will throw an error like
```
error: invalid operands to binary expression ('const value_type' (aka 'const ModuleBase::Vector3<double>') and 'Vector3<int>')
```
However, int type should be allowed to make calculations with double type data and return double results.

### Describe the solution you'd like

Add support for some basic calculations between ModuleBase::Vector3<int> and ModuleBase::Vector3<double>.

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

**Comment by mohanchen** [comment]
make sense
